fixes for bug 471, 472, 469

rmothilal opened the pull request:
this is for the ffg: 
- failing consumer when not found
- validation of fsp with better error messages
- committed fulfilment updating payeeFsp position

rmothilal commented:
@mdebarros it issue is that we cannot commit the message to the consumer because it does not exist anymore. Do you suggest we recreate the consumer and commit the message? I attempted to commit the message but that failed.

elnyry commented on src/models/transfer/facade.js:
```diff
@@ -273,7 +277,7 @@ const saveTransferPrepared = async (payload, stateReason = null, hasPassedValida
     }
 
-    const state = ((hasPassedValidation) ? Enum.TransferState.RECEIVED_PREPARE : Enum.TransferState.INVALID)
+    const state = ((hasPassedValidation) ? Enum.TransferState.RECEIVED_PREPARE : Enum.TransferState.ABORTED)
```
Regarding this one - I think the intention was to use INVALID internally to easily identify requests for which validation failed. Here is the sequence diagram - https://github.com/mojaloop/docs/blob/develop/CentralServices/seq_diagrams/seq-prepare-1.1.1.a.svg

rmothilal replied:
@elnyry Invalid does not exist in the transferState table. This will fail because of a foreign key exception

elnyry replied:
Its in the seeds here - https://github.com/mojaloop/central-ledger/blob/develop/seeds/transferState.js ?

mdebarros commented:
The issue is not that we require a new consumer, it's that we are receiving a message in the wrong topic i.e. consumer prepare for dfsp1 , that consumer the tries to commit the message to the right topic which is for dfsp2.

The correct approach is that we should be logging the error, and then dfsp1 consumer should commit the message.

I.e. the way we look up the consumer is incorrect. We should be looking to the consumer not on the message.value (payload) but rather on the message meta data which hopefully contains the topic from where the message was attained from.

rmothilal commented:
rework to still throw error for missing topic and handle appropriately as well as getting topic from incoming message and not creating it from the fsp. reverted state to invalid for validation errors on prepare. fixed corresponding tests

mdebarros commented on src/handlers/positions/handler.js:
```diff
@@ -204,10 +210,12 @@ const positions = async (error, messages) => {
+        consumer = Kafka.Consumer.getConsumer(kafkaTopic)
+      } catch (e) {
         Logger.info(`no consumer found for topic ${kafkaTopic}`)
+        Logger.info(e)
```
Should be a loger.error.

mdebarros commented on src/handlers/positions/handler.js:
```diff
@@ -181,10 +185,12 @@ const positions = async (error, messages) => {
+          consumer = Kafka.Consumer.getConsumer(kafkaTopic)
+        } catch (e) {
           Logger.info(`no consumer found for topic ${kafkaTopic}`)
+          Logger.info(e)
```
Should be a loger.error.

mdebarros commented on src/handlers/positions/handler.js:
```diff
@@ -137,10 +139,12 @@ const positions = async (error, messages) => {
+        consumer = Kafka.Consumer.getConsumer(kafkaTopic)
+      } catch (e) {
         Logger.info(`no consumer found for topic ${kafkaTopic}`)
+        Logger.info(e)
```
Should be a loger.error.

mdebarros commented on src/handlers/positions/handler.js:
```diff
@@ -84,10 +84,12 @@ const positions = async (error, messages) => {
+        consumer = Kafka.Consumer.getConsumer(kafkaTopic)
+      } catch (e) {
         Logger.info(`no consumer found for topic ${kafkaTopic}`)
+        Logger.info(e)
```
Should be a loger.error.

mdebarros commented on src/handlers/lib/kafka/consumer.js:
```diff
@@ -95,7 +95,7 @@ const getConsumer = (topicName) => {
     return listOfConsumers[topicName].consumer
   } else {
-    return undefined
+    throw Error(`no consumer found for topic ${topicName}`)
```
Should we not capitalise the message?  -> "No consumer....."

mdebarros commented on src/handlers/transfers/handler.js:
```diff
@@ -95,11 +95,14 @@ const prepare = async (error, messages) => {
+      consumer = Kafka.Consumer.getConsumer(kafkaTopic)
+    } catch (e) {
       Logger.info(`no consumer found for topic ${kafkaTopic}`)
+      Logger.info(e)
```
Should be a loger.error.

mdebarros commented on src/handlers/transfers/handler.js:
```diff
@@ -228,10 +231,13 @@ const fulfil = async (error, messages) => {
+      consumer = Kafka.Consumer.getConsumer(kafkaTopic)
+    } catch (e) {
       Logger.info(`no consumer found for topic ${kafkaTopic}`)
+      Logger.info(e)
```
Should be a loger.error.

mdebarros commented on src/handlers/transfers/handler.js:
```diff
@@ -342,13 +348,15 @@ const transfer = async (error, messages) => {
+        consumer = Kafka.Consumer.getConsumer(kafkaTopic)
+      } catch (e) {
         Logger.info(`no consumer found for topic ${kafkaTopic}`)
+        Logger.info(e)
```
Should be a loger.error.

mdebarros commented on src/handlers/transfers/handler.js:
```diff
@@ -359,8 +367,14 @@ const transfer = async (error, messages) => {
+        consumer = Kafka.Consumer.getConsumer(kafkaTopic)
+      } catch (e) {
+        Logger.info(`no consumer found for topic ${kafkaTopic}`)
+        Logger.info()
```
Should be a loger.error.
